From 7e6bb5d9ab6f3f6939579648f32dd92844462d84 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:43:18 -0400
Subject: [infiniband] cxgb4: Only allow kernel db ringing for T4 devs

Message-id: <1410475447-94392-60-git-send-email-svemuri@redhat.com>
Patchwork-id: 93880
O-Subject: [RHEL7.1 PATCH BZ 1124947 059/108] RDMA/cxgb4: Only allow kernel db ringing for T4 devs
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

The whole db drop avoidance stuff is for T4 only.  So we cannot allow
that to be enabled for T5 devices.

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit c2f9da92f2fd6dbf8f40ef4d5e00db688cc0416a)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/qp.c b/drivers/infiniband/hw/cxgb4/qp.c
index f18ef34..086f62f 100644
--- a/drivers/infiniband/hw/cxgb4/qp.c
+++ b/drivers/infiniband/hw/cxgb4/qp.c
@@ -1777,11 +1777,15 @@ int c4iw_ib_modify_qp(struct ib_qp *ibqp, struct ib_qp_attr *attr,
  /*
   * Use SQ_PSN and RQ_PSN to pass in IDX_INC values for
   * ringing the queue db when we're in DB_FULL mode.
+  * Only allow this on T4 devices.
   */
  attrs.sq_db_inc = attr->sq_psn;
  attrs.rq_db_inc = attr->rq_psn;
  mask |= (attr_mask & IB_QP_SQ_PSN) ? C4IW_QP_ATTR_SQ_DB : 0;
  mask |= (attr_mask & IB_QP_RQ_PSN) ? C4IW_QP_ATTR_RQ_DB : 0;
+ if (is_t5(to_c4iw_qp(ibqp)->rhp->rdev.lldi.adapter_type) &&
+     (mask & (C4IW_QP_ATTR_SQ_DB|C4IW_QP_ATTR_RQ_DB)))
+  return -EINVAL;
 
  return c4iw_modify_qp(rhp, qhp, mask, &attrs, 0);
 }
-- 
1.7.1